website: Promote `T.class_of` doc from fragment (#4379)
sorbet/sorbet: website/docs/class-of.md

```

<a href="https://sorbet.run/#%23%20typed%3A%20true%0AT.let(0%2C%20Integer)%0AT.let(1%2C%20Integer)%0AT.let(2%20%2B%202%2C%20Integer)%0A%0AT.let(Integer%2C%20T.class_of(Integer))%0AT.let(42.class%2C%20T.class_of(Integer))">
  → View on sorbet.run
</a>

## `T.class_of` and inheritance

As with [Class Types](class-types.md#inheritance), `T.class_of` types work with
inheritance:

```ruby
# typed: true
extend T::Sig

class Grandparent; end
class Parent < Grandparent; end
class Child < Parent; end

sig {params(x: T.class_of(Parent)).void}
def example(x); end

example(Grandparent)   # error
example(Parent)        # ok
example(Child)         # ok
```

<a href="https://sorbet.run/#%23%20typed%3A%20true%0Aextend%20T%3A%3ASig%0A%0Aclass%20Grandparent%3B%20end%0Aclass%20Parent%20%3C%20Grandparent%3B%20end%0Aclass%20Child%20%3C%20Parent%3B%20end%0A%0Asig%20%7Bparams(x%3A%20T.class_of(Parent)).void%7D%0Adef%20example(x)%3B%20end%0A%0Aexample(Grandparent)%20%20%20%23%20error%0Aexample(Parent)%20%20%20%20%20%20%20%20%23%20ok%0Aexample(Child)%20%20%20%20%20%20%20%20%20%23%20ok">
  → View on sorbet.run
</a>

The most surprising feature of `T.class_of` comes from not understanding

> inheritance in Ruby, especially with `include` or `extend` plus modules.

Author: - `include` and `extend` are not keywords, they're methods.
- I'd prefer not using `/` in written documents like this, because it's ambiguous. `or` is more clear
- We don't use `you` in Sorbet docs, because it leads to a tone like, "you who make mistakes, vs me who is always right" throughout the documentation.

Going to pass on this. Feel free to suggest something else.

  → View on sorbet.run
</a>

Consider we have some code like this:

```ruby
class MyClass
  def some_instance_method; end

>   def self.some_class_methodd; end

Reviewer: Typo
```suggestion
  def self.some_class_method; end
```

end

sig {params(x: T.class_of(MyClass)).void}
def example1(x)
  x.new.some_instance_method  # ok
  x.some_class_method         # ok
end

example1(MyClass)             # ok
```

This is a class which has an instance method and a class method.
`T.class_of(MyClass)` works like we expect.

Now imagine that we have a lot of these classes and we want to factor out an
interface. The straightforward way to do this uses
[`mixes_in_class_methods`](abstract#interfaces-and-the-included-hook), like

> this:

Reviewer 2: Actually I'll retract this suggestion. (The narrative unfolded differently from what I was imagining when I wrote it.)

end

example2(MyClass)             # error: Expected `T.class_of(MyInterface)`
                              #        but found `T.class_of(MyClass)`
```

**These errors are correct**, and we can verify them in the Ruby REPL. First,
let's explain the error on the last line above:

```
❯ MyClass.singleton_class.ancestors
=> [#<Class:MyClass>, MyInterface::ClassMethods, #<Class:Object>, T::Private::Methods::MethodHooks, #<Class:BasicObject>, Class, Module, T::Sig, Object, Kernel, BasicObject]
```

The first two ancestors of the `MyClass` object are itself and
`MyInterface::ClassMethods`. But notably, `#<Class:MyInterface>` **does not**
appear in this list, so Sorbet is correct to say that `MyClass` does not have
type `T.class_of(MyInterface)`. This is because neither `include` nor `extend`
in Ruby will cause `#<Class:MyInterface>` to appear in any ancestors list.

Next, let's explain the other two errors:

```
❯ MyInterface.singleton_class.ancestors
=> [#<Class:MyInterface>, T::Private::MixesInClassMethods, T::Helpers, Module, T::Sig, Object, Kernel, BasicObject]
```

For the `MyInterface` class object, we see that its only ancestor is itself

> (ignoring those common ancestors). Notably, **none** of the classes in this list

Reviewer 2: ```suggestion
(ignoring common ancestors like `Object`). Notably, **none** of the classes in this list
```

> define either a method called `new` (because `Class` is not there) nor

Reviewer 2: ```suggestion
defines either a method called `new` (because `Class` is not there) nor
```

Author: I think this should be `define` (plural) from both my intuition of "does this sound right" and my reading of a handful of grammar articles online.

Reviewer 2: Works for me!

`some_class_method` (because `MyInterface::ClassMethods` is not there).

> While these errors are correct, **we want to be able to type this code**. There

Reviewer 2: ```suggestion
While these errors are technically correct, **we want to be able to type this code**. There
```

are two options:

1.  Use an abstract class instead of an interface.

    Sometimes this is not possible, because the class in question already has a

>     superclass that can't be changed. However, if this is an option to you, it's

Reviewer 2: ```suggestion
    superclass that can't be changed. However, if this option is available, it's
```

    likely the most straightforward. If we change `MyInterface` to
    `MyAbstractClass`, all our problems vanish.

2.  Use `T.all(Class, MyInterface::ClassMethods)`.

>     This is only a partial solution, but might be good enough.

Reviewer 2: ```suggestion
    For our example this is only a partial solution, but in many cases it is good enough.
```

Specifically, option (2) looks like this:

```ruby
sig {params(x: T.all(Class, MyInterface::ClassMethods)).void}
def example3(x)
  x.new.some_instance_method  # error: `some_instance_method` does not exist
  x.some_class_method         # ok
end

example3(MyClass)             # ok
```

> We’re down to only one error now. The error is still technically correct: since
> we’re using `Class` instead of `T.class_of(...)`, Sorbet has no way to know what
> the instance type created by `x.new` will be (it could be anything), so it
> treats the type as `Object`, causing `some_instance_method` to not be found.
> 
> But if this tradeoff is okay, both the top-level call site and the call to
> `x.some_class_method` now typecheck successfully.

Reviewer 2: ```suggestion
We’re down to only one error now. The error is still technically correct: since
we’re using `Class` instead of `T.class_of(...)`, Sorbet has no way to know what
the instance type created by `x.new` will be (it could be anything), so it
treats the type as `Object`, causing `some_instance_method` to not be found.
However, both the top-level call site to `example3` and the call to `x.some_class_method`
now typecheck successfully. In cases where we don't actually need to use instance methods
from `MyInterface`, this may be an acceptable workaround.
```
